### PR TITLE
internal/transport: skip log on EOF when reading client preface

### PR DESCRIPTION
**Pull request by easwars**
Fixes https://github.com/grpc/grpc-go/issues/4234.

RELEASE NOTES: N/A

### CODE REVIEW

File: internal/transport/http2_server.go [hunk 1]
```diff
@@ -266,6 +266,13 @@ func newHTTP2Server(conn net.Conn, config *ServerConfig) (_ ServerTransport, err
...
 	preface := make([]byte, len(clientPreface))
 	if _, err := io.ReadFull(t.conn, preface); err != nil {
+		// In deployments where a gRPC server runs behind a cloud load balancer
+		// which performs regular TCP level health checks, the connection is
+		// closed immediately by the latter. Skipping the error here will help
+		// reduce log clutter.
+		if err == io.EOF {
+			return nil, nil
+		}
 		return nil, connectionErrorf(false, err, "transport: http2Server.HandleStreams failed to receive the preface from client: %v", err)
```
Comment (dfawley) [inline internal/transport/http2_server.go:276]: This seems to only matter for plaintext connections, or else the peer is finishing the handshake.

Is this possibly an incomplete fix?  Where do we handle handshaker errors?  Do they log?

File: internal/transport/http2_server.go [hunk 2]
```diff
@@ -266,6 +266,13 @@ func newHTTP2Server(conn net.Conn, config *ServerConfig) (_ ServerTransport, err
 	// Check the validity of client preface.
 	preface := make([]byte, len(clientPreface))
 	if _, err := io.ReadFull(t.conn, preface); err != nil {
+		// In deployments where a gRPC server runs behind a cloud load balancer
+		// which performs regular TCP level health checks, the connection is
+		// closed immediately by the latter. Skipping the error here will help
+		// reduce log clutter.
+		if err == io.EOF {
```
Comment (dfawley) [inline internal/transport/http2_server.go:273]: What about `io.ErrUnexpectedEOF`?

File: internal/transport/http2_server.go [hunk 3]
```diff
@@ -266,6 +266,13 @@ func newHTTP2Server(conn net.Conn, config *ServerConfig) (_ ServerTransport, err
...
+			return nil, nil
```
Comment (dfawley) [inline internal/transport/http2_server.go:274]: Let's make sure this is documented above and in `NewServerTransport`.  Currently both indicate an error will be returned, or a valid transport.

To make these easier to keep in sync, I'd be happy if we finally delete the current `NewServerTransport` and rename this function to it.  Or use another name (e.g. `NewHTTP2Server[Transport]`) if desired.

Comment (easwars) [reply internal/transport/http2_server.go:279]: Done. Ended up retaining the name `NewServerTransport` to match with the existing `NewClientTransport`.

File: internal/transport/http2_server.go (hunk 1 above)
Comment (dfawley) [reply internal/transport/http2_server.go:281]: > My understanding is that cloud load balancers which perform TCP handshakes do not use TLS

But these logs are from the server, so it doesn't matter what the client is attempting to use.  If it connects without TLS then the server's handshaker will fail it before we get here.

File: internal/transport/http2_server.go (hunk 2 above)
Comment (dfawley) [reply internal/transport/http2_server.go:278]: Is this a regression?  Maybe the logs were happening before that refactor, too?  I would think either error could be safely skipped.

File: internal/transport/http2_server.go (hunk 1 above)
Comment (easwars) [reply internal/transport/http2_server.go:281]: What we could do is that we could handle `io.EOF` as a special case here just like how `err != credentials.ErrConnDispatched` is being handled, and not add a log here.
https://github.com/grpc/grpc-go/blob/master/server.go#L843-L855

File: internal/transport/http2_server.go (hunk 2 above)
Comment (dfawley) [reply internal/transport/http2_server.go:278]: That seems reasonable.

File: internal/transport/http2_server.go (hunk 1 above)
Comment (dfawley) [reply internal/transport/http2_server.go:281]: Yes, something like that might be a good idea, too.

**Review by dfawley — APPROVED** [review]

File: internal/transport/http2_server.go (hunk 1 above)
Comment (easwars) [reply internal/transport/http2_server.go:281]: Done. Made the change to `server.go`.

**Review by dfawley — APPROVED** [review]

**Pull request merged**